modules/file-transfer/src/main/java/org/apache/ignite/internal/network/file/FileTransferServiceImpl.java

```diff
@@ -0,0 +1,402 @@
+                    return (FileProvider<Metadata>) provider;
+                }
+        );
+    }
+
+    @Override
+    public <M extends Metadata> void addFileHandler(
+            Class<M> metadata,
+            FileConsumer<M> handler
+    ) {
+        metadataToHandler.compute(
+                getMessageType(metadata),
+                (k, v) -> {
+                    if (v != null) {
+                        throw new IllegalArgumentException("File handler for metadata " + metadata.getName() + " already exists");
+                    }
+                    return (FileConsumer<Metadata>) handler;
+                }
+        );
+    }
+
+    @Override
+    public CompletableFuture<List<File>> download(String nodeConsistentId, Metadata metadata) {
+        return completedFuture(UUID.randomUUID())
+                .thenApply(transferId -> factory.fileDownloadRequest()
+                        .transferId(transferId)
+                        .metadata(metadata)
+                        .build())
+                .thenCompose(message -> {
+                    Path directory = createTransferDirectory(message.transferId());
+                    FileTransferMessagesHandler handler = fileReceiver.registerTransfer(nodeConsistentId, message.transferId(), directory);
+                    return messagingService.invoke(nodeConsistentId, FILE_TRANSFERRING_CHANNEL, message,
+                                    RESPONSE_TIMEOUT)
+                            .thenApply(FileDownloadResponse.class::cast)
+                            .thenCompose(response -> {
+                                if (response.error() != null) {
+                                    return failedFuture(new FileTransferException(response.error()));
+                                } else {
+                                    return handler.result();
+                                }
+                            });
+                });
+    }
+
+    @Override
+    public CompletableFuture<Void> upload(String nodeConsistentId, Metadata metadata) {
+        return supplyAsync(() -> getFileProvider(metadata))
+                .thenCompose(provider -> provider.files(metadata))
+                .thenCompose(files -> {
+                    FileUploadRequest message = factory.fileUploadRequest()
+                            .metadata(metadata)
+                            .build();
+
+                    if (files.isEmpty()) {
+                        return failedFuture(new FileTransferException("No files to upload"));
+                    } else {
+                        return messagingService.invoke(nodeConsistentId, FILE_TRANSFERRING_CHANNEL, message, RESPONSE_TIMEOUT)
+                                .thenApply(FileUploadResponse.class::cast)
+                                .thenCompose(response -> sendFiles(nodeConsistentId, response.transferId(), files));
+                    }
+                });
+    }
+
+    private static short getMessageType(Class<?> metadata) {
```

Comment:
Maybe we should register handlers by the numeric message types, rather than classes? Cause it looks a bit hacky right now